File: bin/model_performance.py
#!/user/bin/python3


import rpy2.robjects as ro
from rpy2.robjects import pandas2ri
from rpy2.robjects.conversion import localconverter
from rpy2.robjects.packages import importr
import rpy2.robjects as robjects

from pathlib import Path
import os
import scanpy as sc
import numpy as np
import pandas as pd
import anndata as ad
import warnings
import statsmodels.api as sm
from statsmodels.formula.api import ols
from statsmodels.stats.anova import anova_lm
from pathlib import Path
import matplotlib.pyplot as plt
import seaborn as sns
import argparse
import os
import statsmodels as sm
from scipy import stats
import matplotlib.pyplot as plt
from statsmodels.stats.multitest import multipletests
from matplotlib import cm
from matplotlib.colors import Normalize
import statsmodels.api as sm
import statsmodels.formula.api as smf
from statsmodels.formula.api import glm
import itertools
import textwrap
from matplotlib.patches import Patch  
from matplotlib.lines import Line2D

# Function to parse command line arguments
def parse_arguments():
  parser = argparse.ArgumentParser(description="Download model file based on organism, census version, and tree file.")
  parser.add_argument('--weighted_f1_results', type=str, default = "/space/grp/rschwartz/rschwartz/evaluation_summary.nf/SCT_integrated_hsap/aggregated_results/weighted_f1_results.tsv", help="Aggregated weighted results")
  parser.add_argument('--label_f1_results', type=str, default="/space/grp/rschwartz/rschwartz/evaluation_summary.nf/SCT_integrated_hsap/aggregated_results/label_f1_results.tsv", help="Label level f1 results")                                            
  if __name__ == "__main__":
      known_args, _ = parser.parse_known_args()
      return known_args


def run_lmm_random(df, formula, group_var='study'):
    """
    Runs a linear mixed-effects model treating `group_var` as a random effect.
    
    Parameters:
        df (pd.DataFrame): The dataset.
        formula (str): The model formula (e.g., 'f1_score ~ cutoff + variable').
        group_var (str): The grouping variable for the random effect (default: 'study').
    
    Returns:
        pd.DataFrame: Model summary with FDR-corrected p-values.
    """
    # Extract the outcome variable
    outcome_var = formula.split('~')[0].strip()
    
    # Log-transform the outcome (avoid log(0))
    df[outcome_var] = np.log(df[outcome_var] + 1e-6)

    # Fit the model: fixed effects + study as a random effect
    model = smf.mixedlm(formula, df, groups=df[group_var], re_formula="~1").fit()

    # Extract coefficients and p-values
    model_summary_df = model.summary().tables[1]
    
    # Convert to DataFrame
    model_summary_df = pd.DataFrame(model_summary_df.data[1:], columns=model_summary_df.data[0])
    model_summary_df = model_summary_df.set_index(model_summary_df.columns[0])
    model_summary_df["SE"] = model.bse
    
    # Ensure numeric p-values for FDR correction
    model_summary_df["P>|z|"] = pd.to_numeric(model_summary_df["P>|z|"], errors="coerce")
    
    # Apply FDR correction
    pvals = model_summary_df["P>|z|"].dropna()
    _, fdr_corrected_pvals, _, _ = multipletests(pvals, method="fdr_bh")
    model_summary_df.loc[pvals.index, "FDR"] = fdr_corrected_pvals
    
    # Add model fit statistics
    model_summary_df["LogLik"] = model.llf
    model_summary_df["AIC"] = model.aic
    model_summary_df["BIC"] = model.bic
    
    return model, model.summary2().tables[0], model_summary_df


def plot_model_summary(model_summary, outdir, key):
    plt.rcParams.update({'font.size': 30})
    model_summary["FDR<0.05"] = model_summary["FDR"] < 0.05
    model_summary["Term"] = model_summary["Term"].str.replace("T.", "")
    formula = model_summary["formula"].unique()[0]
    
    # Use the subset directly (no need for grouping)
    plt.figure(figsize=(30, max(10, len(model_summary) * 0.5)))  # Scale height dynamically

    ax = sns.barplot(
        data=model_summary, 
        y="Term", 
        x="Coef.", 
        hue="Term",  # Color bars by term
        dodge=False, 
        errorbar=None, 
        palette="tab20"  # Ensures enough colors for multiple terms
    )

    # Highlight significant terms (FDR < 0.05) with black edges or transparency
    for patch, (_, row) in zip(ax.patches, model_summary.iterrows()):
        if row["FDR<0.05"]:
            patch.set_edgecolor("red")  # Add red border for significant terms
            patch.set_linewidth(2)
        else:
            patch.set_alpha(0.5)  # Reduce opacity for non-significant terms

        term_color = patch.get_facecolor()  # Get color of the bar
        y_tick_label = ax.get_yticklabels()[int(patch.get_y() + patch.get_height() / 2)]  # Match tick label to bar
        y_tick_label.set_color(term_color)

    # Set y-axis limits to remove extra space
    ax.set_ylim(-0.5, len(model_summary) - 0.5)        
    for patch, (_, row) in zip(ax.patches, model_summary.iterrows()):
        x_center = patch.get_x() + patch.get_width() / 2  # Get center of the bar horizontally
        y_center = patch.get_y() + patch.get_height() / 2  # Get center of the bar vertically
        plt.errorbar(x=x_center, y=y_center, xerr=2 * row["Std.Err."], fmt="none", color="black", capsize=5, capthick=2)

    # Add a custom legend for FDR < 0.05
    handles, labels = ax.get_legend_handles_labels()

    # Add custom legend for FDR < 0.05
    legend_elements = [
        Line2D([0], [0], marker='o', color='w', markerfacecolor='red', markersize=10, label='FDR < 0.05')
    ]
    ax.legend(handles + legend_elements, labels + ['FDR < 0.05'], loc='upper left', bbox_to_anchor=(1, 1))

    title = f"{key} ({formula})"
    wrapped_title = "\n".join(textwrap.wrap(title, width=40))  # Adjust width as needed

    # Plot
    plt.title(wrapped_title, fontsize=30)      
    plt.xlabel("Coefficient", fontsize=30)
    plt.ylabel("Term", fontsize=30)
    plt.tight_layout()
    plt.savefig(os.path.join(outdir,f"{key}_{formula}_lm_coefficients.png"))


def plot_model_metrics(df_list, formulas):
  plt.rcParams.update({'font.size': 15})
  results = []
    
  for df in df_list:
      for formula in formulas:
          model, model_summary, model_summary_df = run_lmm_random(df, formula)
          for key in df['key'].unique():
              key_df = model_summary_df.copy()
              key_df['Dataset'] = key
              key_df['Formula'] = formula
              results.append(key_df[['Dataset', 'Formula', 'LogLik', 'AIC']])

  # Combine results into a single DataFrame
  results_df = pd.concat(results, ignore_index=True)

  # Plot Rsquared vs. AIC
  plt.figure(figsize=(20, 6))
  scatter = sns.scatterplot(data=results_df, x="AIC", y="Rsquared", hue="Dataset", style="Formula", palette="tab10", edgecolor="black", s=100)
    
  # Add labels
  plt.title("LogLik vs AIC Across Models")
  plt.xlabel("AIC")
  plt.ylabel("LogLik")
    
  # Improve legend
  plt.legend(title="Dataset", bbox_to_anchor=(1, 1))
  plt.grid(True, linestyle="--", alpha=0.6)
  plt.tight_layout()
    
  plt.savefig("model_metrics.png", bbox_inches="tight")
  plt.show()



def run_and_store_model(df, formula, formula_dir, key):
  """Runs the linear model, saves results."""
  model, model_summary, model_summary_coefs = run_lmm_random(df, formula)
  model_summary_coefs["formula"] = formula
  model_summary_coefs["key"] = key
  model_summary_coefs["Term"] = model_summary_coefs.index
  model_summary_coefs.to_csv(os.path.join(formula_dir, "model_summary_coefs.tsv"), sep="\t", index=False)
  model_summary.to_csv(os.path.join(formula_dir, "model_summary.tsv"), sep="\t", index=False)
  plot_model_summary(model_summary_coefs, outdir=formula_dir, key=key)
  return model_summary_coefs
 

# Convert a pandas DataFrame to an R DataFrame
def pandas_to_r(df):
    r_df = ro.pandas2ri.py2rpy(df)
    return r_df
 
 
def main():
    
  args = parse_arguments()
  weighted_f1_results = pd.read_csv(args.weighted_f1_results, sep="\t")
  #weighted_f1_results = weighted_f1_results[weighted_f1_results["cutoff"] == 0]
  organism = weighted_f1_results["organism"].unique()[0]
  #label_results = pd.read_csv(args.label_f1_results, sep="\t")
  #label_results = label_results[label_results["cutoff"] == 0]
  
  
  factor_names = ["study", "reference", "method", "cutoff"]
  #if organism 
  formulas = [
      "weighted_f1 ~ " + " + ".join(factor_names) + " + reference:method + method:cutoff",
      "weighted_f1 ~ " + " + ".join(factor_names) + " + reference:method + method:cutoff +  disease_state",
      "weighted_f1 ~ " + " + ".join(factor_names) + " + reference:method + method:cutoff + disease_state + sex"]
   
  df_list = [group for _, group in weighted_f1_results.groupby('key')]
  # use only "subclass" df
  df_list = [df for df in df_list if df["key"].values[0] == "subclass"]
  plot_model_metrics(df_list, formulas)
  
  model_summary_coefs_combined = [] 
  for df in df_list:
      df["method"] = pd.Categorical(df["method"], categories=["seurat", "scvi"], ordered=True)
      key = df["key"].values[0]
      for formula in formulas:
          formula_dir = os.path.join("weighted_f1", formula.replace(" ",""), key)
          os.makedirs(formula_dir, exist_ok=True)
          model_summary_coefs = run_and_store_model(df, formula, formula_dir, key)
          model_summary_coefs_combined.append(model_summary_coefs)
          
  model_summary_coefs_combined = pd.concat(model_summary_coefs_combined)
  f1_type = "weighted"
  model_summary_coefs_combined.to_csv(f"{f1_type}_model_summary_coefs_combined.tsv", sep="\t", index=False)

      
    
    
      
if __name__ == "__main__":
    main()

File: bin/model_performance.R
library(glmmTMB)
library(dplyr)
library(broom.mixed)
library(multcomp)
library(argparse)
library(dplyr)
library(readr)
library(dplyr)
library(tidyr)
library(ggplot2)
library(gridExtra)
library(stringr)

# Set global theme for background
theme_set(
  theme_minimal(base_size =30 ) +  # Base theme
    theme(
      plot.background = element_rect(fill = "white", color = NA),  # Plot background color
      panel.background = element_rect(fill = "white", color = NA), # Panel background color
      legend.background = element_rect(fill = "white", color = NA) # Legend background color
    )
)

run_beta_model <- function(df, formula, group_var = "study") {
  # Ensure the outcome is within (0,1) for Beta regression
  outcome_var <- all.vars(as.formula(formula))[1]
  df[[outcome_var]] <- pmax(pmin(df[[outcome_var]], 1 - 1e-6), 1e-6)

  random_effect_formula <- paste(formula, "+ (1 |", group_var, ")")
    # Add random effects directly to the formula
  model <- glmmTMB(as.formula(random_effect_formula), data = df, family = beta_family(link = "logit"))

  
  # Extract coefficients and p-values
  summary_df <- tidy(model, effects = "fixed") %>%
    mutate(FDR = p.adjust(p.value, method = "fdr"))
  
  # Model fit statistics
  model_stats <- tibble(
    LogLik = logLik(model),
    AIC = AIC(model),
    BIC = BIC(model)
  )
  
  return(list(model = model, summary = summary_df, stats = model_stats, formula = random_effect_formula))
}


run_and_store_model <- function(df, formula, formula_dir, key) {
  # Run the beta model using the run_beta_model function
  result <- run_beta_model(df, formula, group_var = "study")  # Adjust group_var as needed
  
  # Extract model summary coefficients and add additional info
  model_summary_coefs <- result$summary
  model_summary_coefs$formula <- formula
  model_summary_coefs$key <- key
  model_summary_coefs$LogLik <- result$stats$LogLik
  model_summary_coefs$AIC <- result$stats$AIC
  model_summary_coefs$BIC <- result$stats$BIC

  # Save the model summary and coefficients summary to files
  write.table(model_summary_coefs, file = file.path(formula_dir, paste0(key,"_model_summary_coefs_combined.tsv")), sep = "\t", row.names = FALSE)
 # write.table(result$stats, file = file.path(formula_dir, paste0(key,"_model_stats_.tsv")), sep = "\t", row.names = FALSE)
  
  # Plot the model summary (assuming plot_model_summary is defined)
  plot_model_summary(model_summary = model_summary_coefs, outdir = formula_dir, key = key)
  
  # Return the model summary coefficients
  return(model_summary_coefs)
}

# Function to plot model metrics
plot_model_metrics <- function(df_list, formulas) {
  # Initialize an empty list to store results
  results <- list()
  
  # Iterate over the data frames in df_list and formulas
  for (df in df_list) {
    for (formula in formulas) {
      # Run the model (replace with your function to fit the model)
      model_output <- run_beta_model(df, formula)
      stats <- model_output$stats
      # Add key column for each unique 'key' in df
      for (key in unique(df$key)) {
        key_df <- stats %>% 
          mutate(key = key) %>% 
          mutate(formula = model_output$formula)
        
        results <- append(results, list(key_df))
      }
    }
  }
  # Combine the results into a single data frame
  results_df <- do.call(rbind, results)
  library(patchwork)
  results_df$formula_wrapped <- factor(str_wrap(results_df$formula, width = 30))  # Adjust width as needed
  # Plot AIC vs LogLik
  p <- ggplot(results_df, aes(x = AIC, y = LogLik, color = formula_wrapped, shape = key)) +
  geom_point(size = 10, stroke = 2, position = position_jitter(width = 1, height = 1)) +
    labs(
      title = "AIC vs LogLik Across Models",
      x = "AIC",
      y = "LogLik",
      color = "formula",
      shape = "key"
    ) +
    theme(
      legend.title = element_text(size = 30),
      legend.text = element_text(size = 15),
      plot.title = element_text(size = 30),
      axis.title = element_text(size = 30),
    ) +
   # scale_shape_manual(values = c(16, 17, 18)) +  # Customize shapes if needed
  #  scale_color_brewer(palette = "Set2") +  # Customize colors
    theme(legend.position = "right") + plot_annotation()
   # guides(color = guide_legend(title = "key"), shape = guide_legend(title = "Formula"))
    

  # Save the plot
  ggsave("model_metrics.png", p, width = 20, height = 15, dpi = 250)
}


plot_model_summary <- function(model_summary, outdir, key) {
  # Add an FDR < 0.05 column
  model_summary$`FDR < 0.01` <- model_summary$FDR < 0.01
  model_summary$`FDR < 0.05` <- model_summary$FDR < 0.05
  #formula <- unique(model_summary$formula)
  formula <- unique(model_summary$formula)
  formula_wrapped <- unique(factor(str_wrap(formula, width = 10)))  # Adjust width as needed

  model_summary$term <- model_summary$term %>% gsub("reference"," ", .) %>%
                                          gsub("method"," ", .)
  model_summary <- model_summary[order(model_summary$estimate, decreasing=TRUE), ]
  # Create the base plot
  p <- ggplot(model_summary, aes(x = estimate, y = reorder(term, estimate), fill = `FDR < 0.01`)) +
    geom_bar(stat = "identity", show.legend = TRUE) +
    theme(
      axis.text.y = element_text(color = "black"),  # Y-axis text
      axis.text.x = element_text(color = "black"),  # X-axis text
      axis.title = element_text(size = 25),        # Axis title size
      plot.title = element_text(size = 25, hjust = 0.5),  # Title size
      plot.subtitle = element_text(size = 25),      # Subtitle size
      plot.caption = element_text(size = 25),       # Caption size
      axis.ticks = element_line(size = 1),          # Axis tick size
      legend.text = element_text(size = 25),        # Legend text size
      legend.title = element_text(size = 25)        # Legend title size
    ) +
    labs(title = paste(key, " - ", formula_wrapped),
         x = "Coefficient",
         y = "Term")
        
  
  # Highlight significant terms (FDR < 0.05)
  #p <- p + geom_bar(data = model_summary[model_summary$FDR_05, ], 
                     #stat = "identity", 
                     #color = "red", 
                     #size = 1.5,
                     #show.legend = FALSE) +
    #geom_bar(data = model_summary[!model_summary$FDR_05, ], 
             #stat = "identity", 
             #color = "lightgray", 
              #size = 1.5,
             #show.legend = FALSE)
  
  # Add error bars
  p <- p + geom_errorbar(aes(xmin = estimate - 2 * std.error, xmax = estimate + 2 * std.error), width = 0.2)

  # Save the plot to the specified directory
  ggsave(filename = file.path(outdir, paste0(key, "_", formula, "_lm_coefficients.png")), 
  plot = p, width = 35, height = max(10, nrow(model_summary) * 0.5))
}



parser <- argparse::ArgumentParser()
parser$add_argument("--weighted_f1_results", help = "Path to the weighted_f1_results file", default="/space/grp/rschwartz/rschwartz/evaluation_summary.nf/SCT_integrated_hsap/aggregated_results/weighted_f1_results.tsv")
args <- parser$parse_args()


# Reading the weighted_f1_results file
weighted_f1_results <- read.table(args$weighted_f1_results, sep="\t", header=TRUE)
# fill NA with none
weighted_f1_results[is.na(weighted_f1_results)] <- "None"
# Extract organism (assuming only one unique value in the 'organism' column)
organism <- unique(weighted_f1_results$organism)[1]

# Defining factor names
factor_names <- c("study", "reference", "method", "cutoff")

if (organism == "homo_sapiens") {
  # Defining the formulas
  formulas <- list(
    paste("weighted_f1 ~", paste(c(factor_names, "reference:method", "method:cutoff"), collapse = " + ")),
    paste("weighted_f1 ~", paste(c(factor_names, "reference:method", "method:cutoff", "disease_state"), collapse = " + ")),
    paste("weighted_f1 ~", paste(c(factor_names, "reference:method", "method:cutoff", "disease_state", "sex"), collapse = " + "))
  )
} else if (organism == "mus_musculus") {
  formulas <- list(
    paste("weighted_f1 ~", paste(c(factor_names, "reference:method", "method:cutoff"), collapse = " + ")),
    paste("weighted_f1 ~", paste(c(factor_names, "reference:method", "method:cutoff", "treatment"), collapse = " + ")),
    paste("weighted_f1 ~", paste(c(factor_names, "reference:method", "method:cutoff", "treatment", "genotype"), collapse = " + ")),
    paste("weighted_f1 ~", paste(c(factor_names, "reference:method", "method:cutoff", "treatment", "genotype", "strain"), collapse = " + ")),
    paste("weighted_f1 ~", paste(c(factor_names, "reference:method", "method:cutoff", "treatment", "genotype", "strain","age"), collapse = " + "))
  )
}
# Grouping the data by 'key' column and creating a list of data frames
df_list <- split(weighted_f1_results, weighted_f1_results$key)

plot_model_metrics(df_list, formulas)

formula <- formulas[[1]]

for (df in df_list) {
  for (formula in formulas) {
    formula_dir <- formula %>% gsub(" ", "_", .)
    dir.create(formula_dir, showWarnings = FALSE,recursive=TRUE)
    df$method <- factor(df$method, levels=c("seurat","scvi"))
    # set baseline reference
    if (organism == "homo_sapiens") {
      df$reference <- factor(df$reference)
        all_levels <- levels(df$reference)
        df$reference <- factor(df$reference, levels = c("Human Multiple Cortical Areas SMART-seq", 
                                  all_levels[!all_levels %in% "Human Multiple Cortical Areas SMART-seq"]))

    } # need to set default levels for mmus
    if (organism == "mus_musculus") {
      df$reference <- factor(df$reference)
      all_levels <- levels(df$reference)
      ref_ref <- "An integrated transcriptomic and epigenomic atlas of mouse primary motor cortex cell types"
      df$reference <- factor(df$reference, levels = c(ref_ref, 
                                  all_levels[!all_levels %in% ref_ref]))
      df$study <- factor(df$study)
      study_levels <- levels(df$study)
      study_ref <- "GSE214244"
      df$study <- factor(df$study, levels = c(study_ref,
                                              study_levels[!study_levels %in% study_ref]))

    }

    run_and_store_model(df, formula, formula_dir = formula_dir, key = df$key[1])
  }
}
